### PR TITLE
[7.x] Fix null value injected from container in routes

### LINKED ISSUE #2114: Null parameter from container is removed when injected in controller.
- Laravel Version: 7.0.2
- PHP Version: 7.4.2

### Description:
Hi, 
I would like to inject a service or null in a controller method.

For now, a null value coming from the container is not spliced in the method parameters. So the next argument value are provided to it, or ArgumentCountError is thrown.

It's coming from `\Illuminate\Routing\RouteDependencyResolverTrait::resolveMethodDependencies`:

https://github.com/laravel/framework/blob/1c5caccad9dc49c9233610d1c92a5a7e46ecaeba/src/Illuminate/Routing/RouteDependencyResolverTrait.php#L49-L56

In my opinion, we should accept null value as standard value, whether they have default value or not, the type-hint will raise an error in the method if the type is not ok.

See the actual errors: 
- https://flareapp.io/share/bP9OdL5Q
- https://flareapp.io/share/omwgpomE

### Steps To Reproduce:
```php
class AppServiceProvider extends ServiceProvider
{
    public function register()
    {
        $this->app->bind(Service::class, fn() => null);
    }

    public function boot()
    {
        Route::get('/foo/{id}', function (?Service $service, string $id) {
            // Argument 1 passed ... must be an instance of Service or null, string given.
        });

        Route::get('/not-working', function (?Service $service) {
            // Too few arguments ... 0 passed ... and exactly 1 expected.
        });

        Route::get('/working', function (?Service $service = null) {
            // Working
        });
    }
}
```

I can propose a pull request with pleasure if you want to.

Thanks,
Matt'

### PR DESCRIPTION
**Pull request by mathieutu**
Hi, 

This PR fixes https://github.com/laravel/ideas/issues/2114 (no idea why this bug have been moved to ideas, maybe I haven't made me understood 😕)

Context:
When injecting something witch can be null, the `\Illuminate\Routing\RouteDependencyResolverTrait::resolveMethodDependencies` method doesn't add it to parameters to inject, because it can't difference it from void return.

I've added a default return value, to indicate that we want to strip the value, to allow null value.

- This PR should not be a breaking change: all tests pass, and the previous behavior was not changed.
- A test to address this case was added.

You can reproduce the bug with:
```php
class AppServiceProvider extends ServiceProvider
{
    public function register()
    {
        $this->app->bind(Service::class, fn() => null);
    }

    public function boot()
    {
        Route::get('/foo/{id}', function (?Service $service, string $id) {
            // Argument 1 passed ... must be an instance of Service or null, string given.
        });

        Route::get('/not-working', function (?Service $service) {
            // Too few arguments ... 0 passed ... and exactly 1 expected.
        });

        Route::get('/working', function (?Service $service = null) {
            // Working
        });
    }
}
```

And see the errors thrown:
- https://flareapp.io/share/bP9OdL5Q
- https://flareapp.io/share/omwgpomE

Thanks.

Matt'

PS: If you read this, and can answer to this https://twitter.com/mathieutu/status/1236997027122249728?s=20, I would like to propose a PR for the router with a quite important breaking change, and so understand why the behavior of Laravel routing is like that, and why it is different from _standard_ one.
Thanks.

### CODE REVIEW

**Comment by mathieutu** [comment]
StyleCI's not happy, but I'm not comfortable with the change proposed, as it makes the function not understandable.

```diff
-        $container->bind(RoutingTestUserModel::class, function() {
-            return null;
+        $container->bind(RoutingTestUserModel::class, function () {
         });
```

**Comment by GrahamCampbell** [comment]
Laravel's code standard is like this because when a fucntion that has no return statement is called and the result is used or assigned to a variable, it is the same as using the value `null` in its place. The actual return statement can be optimized away.

**Comment by deleugpn** [comment]
I actually tried to fix this about 2 weeks ago and my patch was so bad that I never sent it. This looks like a really good patch with minor code changes to fix a moderate problem.
As someone who failed to attempt this, I can appreciate the nice job you did there.

**Comment by mpyw** [comment]
Related:

- [[5.6] Allow nullable method injection by mpyw · Pull Request #22488 · laravel/framework](https://github.com/laravel/framework/pull/22488)

File: src/Illuminate/Routing/RouteDependencyResolverTrait.php [hunk 1]
```diff
@@ -41,12 +41,12 @@ public function resolveMethodDependencies(array $parameters, ReflectionFunctionA
 
         $values = array_values($parameters);
 
+        $shouldSkipValue = new \StdClass;
```
Comment (mpyw) [inline src/Illuminate/Routing/RouteDependencyResolverTrait.php]: Isn't its real name `stdClass`?

```suggestion
        $shouldSkipValue = new \stdClass;
```

**Pull request merged**